YNU-864: Learn introduction rewrite (v1 mental model) (#152)
layer-3/docs: docs/nitrolite/learn/introduction/architecture-at-a-glance.mdx

---
sidebar_position: 2
title: Architecture at a Glance
description: High-level overview of Yellow Network's v1 architecture
keywords: [architecture, state channels, Nitrolite, Nitronode, ChannelHub, smart contracts]
---

# Architecture at a Glance

> Yellow Network v1 combines application code, SDK clients, Nitronode coordination, and ChannelHub enforcement so most updates happen off-chain while the latest mutually signed state remains enforceable on-chain.

Reviewer: Let's remove "Nitronode coordination", so that we don't confuse users into thinking that there are several Nodes.

Author: Removed the Nitronode coordination phrase in [2c41788](https://github.com/layer-3/docs/commit/2c41788c38c77d8bb8cd978b7f9b06e4964a4ee7).

| **Nitronode** | Validates transitions, signs channel states, coordinates off-chain updates | Sub-second in normal operation | No gas |
| **Blockchain layer** | ChannelHub state enforcement, locked funds, challenge and close paths | Block time | Gas fees |

---

## ChannelHub

ChannelHub is the v1 on-chain entrypoint for channel enforcement. It works with ChannelEngine, EscrowDepositEngine, EscrowWithdrawalEngine, and signature validators. Builders interact with ChannelHub as the contract that creates channels, processes deposits and withdrawals, checkpoints signed states, accepts challenges, and closes channels.

ChannelHub validates:

- channel definitions and channel identifiers
- approved signature validation modes
- state versions and intents
- home-ledger and non-home-ledger invariants
- locked-funds accounting
- challenge and close status transitions

Routine transfers and app-session updates are signed off-chain. The blockchain is used for create, deposit, withdraw, checkpoint, challenge, close, escrow, and migration operations.

---

## Off-chain layer

The off-chain layer handles frequent state advancement without blockchain transactions.

### Nitronode

> Nitronode is operated by independent node operators using open-source software developed and maintained by Layer3 Fintech Ltd. It:

Reviewer: ```suggestion
Nitronode is operated by an independent node operator using open-source software developed and maintained by Layer3 Fintech Ltd. It:
```

Author: Applied this wording in [2c41788](https://github.com/layer-3/docs/commit/2c41788c38c77d8bb8cd978b7f9b06e4964a4ee7).

- validates v1 channel transitions
- signs mutually agreed channel states
- issues pending receive and release states for user acknowledgement
- coordinates cross-chain escrow and app-session flows
- serves v1 RPC methods over WebSocket

### v1 RPC

v1 RPC is the wire surface used by SDKs and low-level clients. Builders should prefer the SDK unless they are implementing an SDK, proxy, debugger, or language integration.

---

## How Funds Flow

Funds are represented by channel states with a home ledger and, when needed, a non-home ledger.

```mermaid
flowchart TB
  wallet["User wallet"]
  hub["ChannelHub"]
  home["Home channel<br/>home ledger"]
  nonhome["Non-home ledger<br/>escrow in progress"]
  session["App session<br/>extension state"]

  wallet -->|"approve + home deposit"| hub
  hub -->|"locked funds"| home
  home <-->|"transfer / acknowledge"| home
  home -->|"commit"| session
  session -->|"release"| home
  home <-->|"escrow initiate / finalize"| nonhome
  home -->|"home withdrawal / close"| wallet
```

| Concept | What it means |
|-------|---------------|
| **Home channel** | The user and Node channel for a unified asset. |

> | **Home ledger** | The authoritative ledger for the enforcement chain. |

Reviewer: Let's state this explicitly.

```suggestion
| **Home ledger** | The authoritative single-token ledger for the enforcement chain. |
```

Author: Applied the home-ledger wording in [2c41788](https://github.com/layer-3/docs/commit/2c41788c38c77d8bb8cd978b7f9b06e4964a4ee7).

> | **Non-home ledger** | A temporary ledger for cross-chain escrow, withdrawal, or migration. |

Reviewer: ```suggestion
| **Non-home ledger** | A temporary single-token ledger for cross-chain escrow, withdrawal, or migration. |
```

Author: Applied the non-home-ledger wording in [2c41788](https://github.com/layer-3/docs/commit/2c41788c38c77d8bb8cd978b7f9b06e4964a4ee7).

| **App session** | An extension state funded by commit transitions and settled by release transitions. |

---

## Channel Lifecycle

> The v1 RPC exposes channel status as `void`, `open`, `challenged`, and `closed`.

Reviewer: Currently we also have `closing`, but most probably will also get a new statuses in the current round of audit.
Therefore, either this PR should hang for a while, or a follow-up should be created.

Author: Added `closing` and adjusted the lifecycle wording in [2c41788](https://github.com/layer-3/docs/commit/2c41788c38c77d8bb8cd978b7f9b06e4964a4ee7).

```mermaid
stateDiagram-v2
    [*] --> VOID
    VOID --> OPEN: create or deposit
    OPEN --> OPEN: signed off-chain updates
    OPEN --> OPEN: checkpoint higher-version state
    OPEN --> CHALLENGED: challenge
    CHALLENGED --> OPEN: higher-version response

>     CHALLENGED --> CLOSED: challenge expires and close executes

Reviewer: The label "challenge expires and close executes" implies the close happens automatically on expiry. The protocol docs (enforcement-and-settlement.mdx) say explicitly: "After the challenge period expires without resolution, the disputed state becomes final. A separate close call is still required to release locked funds."

**Suggestion:** Change the label to `CHALLENGED --> CLOSED: challenge expires; close required` to signal that one more call is needed.

Author: Updated the lifecycle label so it shows that close is still required in [2c41788](https://github.com/layer-3/docs/commit/2c41788c38c77d8bb8cd978b7f9b06e4964a4ee7).
